### PR TITLE
Addresses #82 - Add raw and rgb screenshot functionality

### PR DESCRIPTION
**Pull request by tischi**

### CODE REVIEW

File: src/main/java/sc/fiji/bdvpg/bdv/BdvUtils.java [hunk 1]
```diff
@@ -298,4 +297,11 @@ public static ARGBType getSourceColor(BdvHandle bdvHandle, int sourceId )
         return new double[]{ displayRangeMin, displayRangeMax };
     }
 
+	public static Converter< ?, ARGBType > getConverter( Bdv bdv, int sourceIndex )
```
Comment (NicoKiaru) [inline src/main/java/sc/fiji/bdvpg/bdv/BdvUtils.java:321]: This is supposed to disappear (indexing logic). Can't you use directly SourceAndConverter instead ? There's a way to retrieve all SourceAndConverters from a Bdv window with the SourceAndConverterDisplayService (https://github.com/bigdataviewer/bigdataviewer-playground/blob/bf0e03687828d70d23822287937d719911d2c4ff/src/main/java/sc/fiji/bdvpg/scijava/services/SourceAndConverterBdvDisplayService.java#L584), and also there's a function to know if it's visible (https://github.com/bigdataviewer/bigdataviewer-playground/blob/bf0e03687828d70d23822287937d719911d2c4ff/src/main/java/sc/fiji/bdvpg/scijava/services/SourceAndConverterBdvDisplayService.java#L160)

**Review by NicoKiaru — COMMENTED** [review]
I don't grasp everything that' happening here. WHen you have time, it would be nice to have comments in ScreenSHotMaker

**Comment by NicoKiaru** [comment]
You can resolve the conflicts and then I can merge it

File: src/main/java/sc/fiji/bdvpg/bdv/BdvUtils.java (hunk 1 above)
Comment (tischi) [reply src/main/java/sc/fiji/bdvpg/bdv/BdvUtils.java:321]: done. thanks for the suggestions!

**Comment by tischi** [comment]
@NicoKiaru I would like to add the ScreenShot functionality somewhere where it is easy to access.

1. Should I add it to the PopupMenu? I do not manage, because I think only actions that consume `sac`s are welcome there.

2. Should I add it as a keyboard shortcut to the `BdvCreator`? Which shortcut would you suggest? `C` is already taken now... maybe `shift S` for **S**creenshot? `S` is already defined within Bdv.

**Comment by tischi** [comment]
@NicoKiaru Can you maybe help?
When running the `ScreenshotDemo` it only find one of the two sources visible. I do not understand why and have trouble debugging your code for that. 
Could you help me by debugging here: https://github.com/bigdataviewer/bigdataviewer-playground/pull/91/commits/23453569e0e3a59c0d3625baad9c255f7ae624d5 ?
As said, I think it should evaluate **both** `sac`s to `true`

**Comment by NicoKiaru** [comment]
> 
> 
> @NicoKiaru Can you maybe help?
> When running the `ScreenshotDemo` it only find one of the two sources visible. I do not understand why and have trouble debugging your code for that.
> Could you help me by debugging here: [2345356](https://github.com/bigdataviewer/bigdataviewer-playground/commit/23453569e0e3a59c0d3625baad9c255f7ae624d5) ?
> As said, I think it should evaluate **both** `sac`s to `true`

So I hope the isvisible issue should be resolved

**Comment by NicoKiaru** [comment]
Ok, let's merge this, @tischi  ? 

An enhancement would be to use the same projector as the bdv window. because now it ignores the occluding part, for instance.

I could open an enhancement issue or if you feel this is simple you can do it directly before merging ?

**Comment by tischi** [comment]
@NicoKiaru 
I had a look; some work, but doable before merging. However we would need to be able to know which AccumulateProjectionFactory is used for a given bdvHandle. Do you have an idea how to do this best?

**Comment by tischi** [comment]
@NicoKiaru 
Ok, works now. Required more changes than expected. If you run the ScreenShotDemo you could try it out. 

Question: How do we make the Screenshot accessible to the user?  **Should we add a Keyboard shortcut for now**? And maybe later add it to the context menu?

**Comment by NicoKiaru** [comment]
Cool! I'll try it.

If you want to add the shortcut to BdvCreator for now that's fine.

We'll have to modularize this later when needed.

**Comment by tischi** [comment]
To add this to `BdvCreator` I need the `CommandService` within `BdvCreator`, because I need to run `ScreenShotMakerCommand`. How should I do this? Can I get it from somewhere?

**Comment by NicoKiaru** [comment]
Can't you run the action directly in 
https://github.com/bigdataviewer/bigdataviewer-playground/blob/64bee42b70f731190e5c628887aefa401436b179/src/main/java/sc/fiji/bdvpg/bdv/BdvCreator.java#L72

and skipping the commandservice ?

**Comment by tischi** [comment]
But there should be the Command UI popping up, for this I need `ij.command().run( ScreenShotMakerCommand.class )` or not?

**Comment by tischi** [comment]
The `SourceAndConverterBdvDisplayService` has a `CommandService`.
We could add a `get` method. It does not feel very logical though...

Other option: we add a `CommandService` as another service to `SourceAndConverterServices`, like `SourceAndConverterServices.getCommandService()`.

**Comment by NicoKiaru** [comment]
you can have a look at the former commit

* I had to tweak the 'actions' registering in the service, because they are supposed to get sourceandconverters as inputs. 

I think we could avoid this and make screenshot a bit more elegant:
* If it's launched with the key from a bdv window : send all visible sources by default
* If it's launched from a FIJI menu : let the user select everything or just a subset of the sources

This last point needs a bit of tweaking if we want the user to have displayed only the sources present in the bdv window

We can also let it like that

**Comment by tischi** [comment]
@NicoKiaru did you look into the bug yet? 
```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at sc.fiji.bdvpg.bdv.BdvCreator.lambda$addSourceAndConverterContextMenuBehaviour$0(BdvCreator.java:84)
	at org.scijava.ui.behaviour.MouseAndKeyHandler.handleKeyPressed(MouseAndKeyHandler.java:596)
	at org.scijava.ui.behaviour.MouseAndKeyHandler.keyPressed(MouseAndKeyHandler.java:504)
```

**Comment by NicoKiaru** [comment]
DId you retry ? Do a 'clean'

I cannot reproduce the bug ...

**Comment by tischi** [comment]
Works now also for me. Merging.

**Pull request merged**